Surface Pressure Stall Information (PSI) metrics

roycaihw opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Surface [PSI metrics from cadvisor](https://github.com/google/cadvisor/pull/3649) to [node summary API](https://kubernetes.io/docs/reference/instrumentation/node-metrics/). [Manual test result](https://gist.github.com/roycaihw/426c8c6de3fc658c7e029cb53d2ee34c). 

This implements the first phase of [this KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4205-psi-metric#phase-1). Ref https://github.com/kubernetes/enhancements/issues/4205

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Add Pressure Stall Information (PSI) metrics to node metrics.
```

/sig node
/cc @haircommander @ndixita

kannon92 commented:
Just a quick glance. For alpha features I’d expect to see any feature related code guarded by a feature gate. So when the gate is off we should not emit PSI metrics.

haircommander commented on pkg/kubelet/stats/cadvisor_stats_provider.go:
```diff
@@ -156,6 +156,7 @@ func (p *cadvisorStatsProvider) ListPodStats(ctx context.Context) ([]statsapi.Po
 			podStats.CPU = cpu
 			podStats.Memory = memory
 			podStats.Swap = cadvisorInfoToSwapStats(podInfo)
+			podStats.IO = cadvisorInfoToIOStats(podInfo)
```
can you also extend the cri_stats_provider and cri-api to have the structs needed for the changes to the summmary api?

roycaihw replied:
@haircommander Done. PTAL

roycaihw commented:
Got a [green e2e run](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/130701/pull-kubernetes-node-e2e-containerd/1901519011359756288). Now adding the feature gate commit (which will move the e2e test to the alpha test suite)

roycaihw commented:
> For alpha features I’d expect to see any feature related code guarded by a feature gate. So when the gate is off we should not emit PSI metrics.

@kannon92 Good point. I added an alpha feature gate. PTAL

haircommander commented on pkg/kubelet/stats/cri_stats_provider.go:
```diff
@@ -583,6 +590,30 @@ func (p *criStatsProvider) addSwapStats(
+	if !utilfeature.DefaultFeatureGate.Enabled(features.KubeletPSI) {
+		return
+	}
+	// try get IO stats from cadvisor first.
+	podCgroupInfo := getCadvisorPodInfoFromPodUID(podUID, allInfos)
+	if podCgroupInfo != nil {
+		ps.IO = cadvisorInfoToIOStats(podCgroupInfo)
+		return
+	}
+
+	if cs.IO != nil {
+		if ps.IO == nil {
+			ps.IO = &statsapi.IOStats{Time: cs.IO.Time}
+		}
+		// Pod level PSI stats cannot be calculated from container level
+	}
```
I think these should be reversed

roycaihw replied:
Could you elaborate why? I was following the suit from other methods like `addPodCPUMemoryStats` and `addSwapStats`

haircommander replied:
ah okay that's fine for now then

haircommander commented on pkg/kubelet/stats/cri_stats_provider.go:
```diff
@@ -592,6 +623,33 @@ func (p *criStatsProvider) addProcessStats(
+	if stats == nil {
+		return nil
+	}
+	result := &cadvisorapi.PSIStats{}
```
more of what I was suggesting is we update the CRI API to collect PSI stats instead of always falling back to cadvisor

haircommander replied:
oh I see sorry I didn't understand we're converting from cri stats to cadvisor

haircommander commented on pkg/kubelet/stats/cri_stats_provider.go:
```diff
@@ -592,6 +623,33 @@ func (p *criStatsProvider) addProcessStats(
 	ps.ProcessStats = mergeProcessStats(ps.ProcessStats, processStats)
 }
 
+func makePSIStats(stats *runtimeapi.PsiStats) *cadvisorapi.PSIStats {
```
nit: can this be defined below where it's called?

bart0sh commented:
/triage accepted

haircommander commented:
one more thing: are we exposing the PSI prometheus metrics coming from cadvisor ? I think we need to enable it when we initialize the handler https://github.com/kubernetes/kubernetes/blob/73f54b67b29d77601b0bd42ad8b4992925b9df47/pkg/kubelet/server/server.go#L456

roycaihw commented:
https://github.com/kubernetes/kubernetes/pull/130701#issuecomment-2736734088 @haircommander Nice catch. I updated and squashed the commits. Here is the [diff](https://gist.github.com/roycaihw/0d0481d0321e7f5fa06f71dd9031a0a8). The prometheus metrics are also behind the alpha feature gate.

ndixita commented on pkg/kubelet/stats/cadvisor_stats_provider.go:
```diff
@@ -156,6 +156,9 @@ func (p *cadvisorStatsProvider) ListPodStats(ctx context.Context) ([]statsapi.Po
 			podStats.Memory = memory
 			podStats.Swap = cadvisorInfoToSwapStats(podInfo)
+			if utilfeature.DefaultFeatureGate.Enabled(features.KubeletPSI) {
+				podStats.IO = cadvisorInfoToIOStats(podInfo)
```
Do we want to update the unit tests for changes inside stats/?

roycaihw replied:
@ndixita Added. PTAL

ndixita replied:
Thank you

ndixita commented:
Is there a presubmit job that has KubeletPSI enabled to run the e2e with changes in this PR? If not, let's add one and run that presubmit job on this PR. example: https://github.com/kubernetes/test-infra/pull/33767/files

roycaihw commented:
https://github.com/kubernetes/kubernetes/pull/130701#issuecomment-2737725320

> Is there a presubmit job that has KubeletPSI enabled to run the e2e with changes in this PR?

Yes I think [pull-kubernetes-node-e2e-containerd-alpha-features](https://github.com/kubernetes/test-infra/blob/adde8e237a7c98ac16e945d6d493cf28f2e61708/config/jobs/kubernetes/sig-node/sig-node-presubmit.yaml#L474C11-L527) enables all alpha feature gates and runs the test. I'm adding the feature tag to the e2e test so it can be picked up by the [focus filter](https://github.com/kubernetes/test-infra/blob/adde8e237a7c98ac16e945d6d493cf28f2e61708/config/jobs/kubernetes/sig-node/sig-node-presubmit.yaml#L516)

kannon92 commented:
> [#130701 (comment)](https://github.com/kubernetes/kubernetes/pull/130701#issuecomment-2737725320)
> 
> > Is there a presubmit job that has KubeletPSI enabled to run the e2e with changes in this PR?
> 
> Yes I think [pull-kubernetes-node-e2e-containerd-alpha-features](https://github.com/kubernetes/test-infra/blob/adde8e237a7c98ac16e945d6d493cf28f2e61708/config/jobs/kubernetes/sig-node/sig-node-presubmit.yaml#L474C11-L527) enables all alpha feature gates and runs the test. I'm adding the feature tag to the e2e test so it can be picked up by the [focus filter](https://github.com/kubernetes/test-infra/blob/adde8e237a7c98ac16e945d6d493cf28f2e61708/config/jobs/kubernetes/sig-node/sig-node-presubmit.yaml#L516)

I do not see you actually use that FeatureGate label on your e2e tests so I think this is not valid.

mrunalp commented on pkg/kubelet/stats/cri_stats_provider_test.go:
```diff
@@ -1191,9 +1191,9 @@ func checkCRIPSIStats(assert *assert.Assertions, want, got *cadvisorapiv1.PSISta
-	assert.Equal(want.Avg10, got.Avg10)
-	assert.Equal(want.Avg60, got.Avg60)
-	assert.Equal(want.Avg300, got.Avg300)
+	assert.InDelta(want.Avg10, got.Avg10, 0.01)
```
Can we fold these changes into the commit where the changes are first introduced?

roycaihw commented:
https://github.com/kubernetes/kubernetes/pull/130701#issuecomment-2741402239
> I do not see you actually use that FeatureGate label on your e2e tests so I think this is not valid.

@kannon92 Yes you're right. See my other comment: https://github.com/kubernetes/kubernetes/pull/130701#issuecomment-2739621627. I plan to have a separate follow-up PR to address this.

roycaihw committed:
Add alpha feature gate KubeletPSI

roycaihw committed:
API definition for PSI metrics

roycaihw committed:
Surface cadvisor PSI metrics to summary API and prometheus metrics

roycaihw committed:
Extend CRI API to support PSI

roycaihw committed:
Extend CRI stats provider to support PSI

roycaihw committed:
Unit and E2E tests

roycaihw committed:
generated

roycaihw commented:
All required tests passed. I'm squashing the commits as suggested.

kannon92 commented:
> The decision should not block this PR, since the alpha feature testing would be part of the CI, not the PR presubmits.

I disagree with this.

So you are saying I am not responsible for testing or verifying this PR. 

If there are any failures in the CI due to this, it could block other people from merging.

Can you confirm at least that one of these presubmits ran the summary_test.go runs? So there is not a failure during code freeze?

roycaihw commented:
https://github.com/kubernetes/kubernetes/pull/130701#issuecomment-2741492300

tl'dr: I added a commit to tag `summary_test.go` with the feature gate (option 1 below). PTAL.

/test pull-kubernetes-node-e2e-containerd-alpha-features

-----

For more context: 

- Currently the presubmit `pull-kubernetes-node-e2e-containerd` runs `summary_test.go`. It's passing with the feature gate disabled (i.e. it doesn't check PSI stats).
- Currently there is no job (presubmit or CI) that runs `summary_test.go` with the feature gate enabled (i.e. checking PSI stats).
  - We had one signal previously that `pull-kubernetes-node-e2e-containerd` passed before I introduced the feature gate, i.e. it passed when it checked PSI stats.

We definitely want jobs that run `summary_test.go` with the feature gate enabled. I think we have a few options:
1. Add the feature gate label to the e2e test (with a caveat that we need to [migrate from ginkgo to f](https://github.com/kubernetes/kubernetes/pull/130701#issuecomment-2739621627)). I'm a little concerned with this approach, because this test is a conformance test. I'm not sure if there would be any push back on tagging conformance tests with feature gates.
2. Add a new job in test-infra that focus on summary_test.go . This way we don't need to add the feature gate label to the test

haircommander commented:
shouldn't there be an alpha e2e job somewhere? I think that could cover us if so

haircommander commented:
yeah I believe since this is a kubelet only feature gate we should be good. the feature is on in the kubelet but not in the apiservers

roycaihw commented:
Yep, `pull-kubernetes-node-e2e-containerd-alpha-features` is the alpha job. 

The job only runs tests [with Feature tagging](https://github.com/kubernetes/test-infra/blob/b9bfeca372742dea635d44ae0c7dcabdff02f25a/config/jobs/kubernetes/sig-node/sig-node-presubmit.yaml#L516) (you can also see [recent runs](https://testgrid.k8s.io/sig-node-presubmits#pr-node-kubelet-containerd-alpha-features) on what tests are triggered). The latest commit is to add the Feature tagging to `summary_test.go` (more context [here](https://github.com/kubernetes/kubernetes/pull/130701#issuecomment-2741568411))

You can also see I [ran the command](https://github.com/kubernetes/kubernetes/pull/130701#issuecomment-2741568411) to manually [trigger the job](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/130701/pull-kubernetes-node-e2e-containerd-alpha-features/1902816917865893888) (it doesn't run on PRs by default, because it's a CI job not a presubmit job).

kannon92 commented:
> You can also see I https://github.com/kubernetes/kubernetes/pull/130701#issuecomment-2741568411 to manually [trigger the job](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/130701/pull-kubernetes-node-e2e-containerd-alpha-features/1902816917865893888) (it doesn't run on PRs by default, because it's a CI job not a presubmit job).

I think we are agreement but using different language..

Technically this is an optional presubmit and triggering this is the right decision.

CI jobs are the periodics that run against main. So typically you want to confirm that your job correctly runs on presubmits before you merge.

If you don't, the CI will show a regression and typically the stance is to revert the PR that caused the failure.

You are doing the right thing in triggering presubmits to test your work!

Thank you for that.

kannon92 commented:
/test pull-kubernetes-node-e2e-containerd-alpha-features

roycaihw commented:
Discussed offline with @kannon92. The alpha feature test failed due to a discrepancy in how feature gate enablement is checked in our test framework. Opening https://github.com/kubernetes/test-infra/pull/34567 to work around. See https://github.com/kubernetes/test-infra/pull/34567 for more details.

roycaihw commented:
/test pull-kubernetes-node-e2e-containerd-alpha-features

roycaihw commented:
The [latest run](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/130701/pull-kubernetes-node-e2e-containerd-alpha-features/1902841107805376512) failed but I think it's unrelated to this PR. 

Reading the result, the PSI stats were correct for [kubelet, runtime and pod](https://github.com/kubernetes/kubernetes/blob/838f3c085261adadfed90d570cc2c65670c73a03/test/e2e_node/summary_test.go#L143-L147). However, it seems some of the busybox containers didn't boot correctly. The summary API not just missed PSI stats, but also missed all pre-existing container stats, like `.Memory.AvailableBytes` and `.Network`. It seems for these containers, the cgroup is empty. 

I suspect some other alpha feature prevented `busybox-container` from running. The issue never got surfaced since we never ran `summary_test.go` with alpha feature gates enabled in the past.

roycaihw commented:
> I suspect some other alpha feature prevented busybox-container from running

I tried to confirm if that's the case. I opened a [test PR](https://github.com/kubernetes/kubernetes/pull/130958/commits/34ef5448cc6efc3586f64996bd4cf1337ab197ff) to enable the feature gate by default. [pull-kubernetes-node-e2e-containerd](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/130958/pull-kubernetes-node-e2e-containerd/1902859770180669440) passed. I think that's great evidence that the alpha test is broken by some other alpha feature. 

Given that, instead of running `summary_test.go` with all alpha features in `pull-kubernetes-node-e2e-containerd-alpha-features`, I'm [creating a new presubmit job](https://github.com/kubernetes/test-infra/pull/34570) to run `summary_test.go` only with `KubeletPSI` enabled. I'm also reverting the Feature tag so that `summary_test.go` is not included in `pull-kubernetes-node-e2e-containerd-alpha-features`

roycaihw commented:
/test pull-kubernetes-node-e2e-containerd-kubelet-psi

roycaihw commented:
/test pull-kubernetes-node-e2e-containerd-kubelet-psi

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1845,6 +1871,13 @@ message MemoryUsage {
+message IoUsage {
+    // Timestamp in nanoseconds at which the information were collected. Must be > 0.
+    int64 timestamp = 1;
+    PsiStats psi = 2;
```
nit: Add comment for field

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1816,6 +1841,7 @@ message CpuUsage {
     // Total CPU usage (sum of all cores) averaged over the sample window.
     // The "core" unit can be interpreted as CPU core-nanoseconds per second.
     UInt64Value usage_nano_cores = 3;
+    PsiStats psi = 4;
```
nit: add field comment

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1807,6 +1811,27 @@ message WindowsContainerStats {
 
+// PSI statistics for an individual resource.
+message PsiStats {
+	// PSI data for all tasks of in the cgroup.
```
nit: `PSI data for all tasks in the cgroup.`

roycaihw commented:
[pull-kubernetes-node-e2e-containerd-kubelet-psi](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/130701/pull-kubernetes-node-e2e-containerd-kubelet-psi/1902878276330721280) passed

kannon92 commented on staging/src/k8s.io/dynamic-resource-allocation/go.sum:
```diff
@@ -55,6 +55,7 @@ github.com/golang/glog v1.2.2/go.mod h1:6AhwSGph0fcJtXVM/PEHPqZlFeoLxhs7/t5UDAwm
 github.com/golang/protobuf v1.5.4 h1:i7eJL8qZTpSEXOPTxNKhASYpMn+8e5Q6AdndVa1dWek=
 github.com/golang/protobuf v1.5.4/go.mod h1:lnTiLA8Wa4RWRcIUkrtSVa5nRhsEGBg48fD6rSs7xps=
 github.com/google/btree v1.1.3/go.mod h1:qOPhT0dTNdNzV6Z/lhRX0YXUafgPLFUh+gZMl761Gm4=
+github.com/google/cadvisor v0.52.1/go.mod h1:OAhPcx1nOm5YwMh/JhpUOMKyv1YKLRtS9KgzWPndHmA=
```
@pohly can you approve this for DRA side?

liggitt replied:
the kubelet API module spreading cadvisor as a dependency is a bit suspect. *populating* the kubelet stats from cadvisor is totally fine, but embedding cadvisor types in the kubelet stats types means all kubelet API consumers now depend on cadvisor, which is… a lot

haircommander replied:
Ooh great catch @liggitt

/hold

haircommander replied:
We could copy it and transfer over

kannon92 reviewed (commented):
Thank you for working on confirmation of the tests!

seeing this pass in CI makes me happy.

/lgtm

roycaihw commented:
Addressed [field documentation comments](https://github.com/kubernetes/kubernetes/pull/130701#discussion_r2006661344) and squashed. 

/test pull-kubernetes-node-e2e-containerd-kubelet-psi

ndixita commented:
/lgtm

ndixita commented:
/hold cancel as the corresponding test job is in

roycaihw commented:
The [hold](https://github.com/kubernetes/kubernetes/pull/130701#issuecomment-2741661410) was for test result, which was [resolved](https://github.com/kubernetes/kubernetes/pull/130701#pullrequestreview-2704455796)

liggitt commented on staging/src/k8s.io/kubelet/pkg/apis/stats/v1alpha1/types.go:
```diff
@@ -225,6 +235,9 @@ type CPUStats struct {
 	UsageCoreNanoSeconds *uint64 `json:"usageCoreNanoSeconds,omitempty"`
+	// CPU PSI stats.
+	// +optional
+	PSI *cadvisorapi.PSIStats `json:"psi,omitempty"`
```
I'm surprised to see us embedding another project's API in our API surface… does this mean anything cadvisor adds we have to support and surface?

mrunalp replied:
Yes, good point. We need to decouple from cadvisor and maintain our own type here.

roycaihw replied:
Done. Thanks for catching this!

mrunalp commented:
/hold 
for the API comment re: cadvisor from @liggitt

roycaihw commented:
/test pull-kubernetes-node-e2e-containerd-kubelet-psi

roycaihw commented:
/test pull-kubernetes-node-e2e-containerd-kubelet-psi

roycaihw commented:
/test pull-kubernetes-node-e2e-containerd-kubelet-psi

haircommander commented:
/unhold
/lgtm

haircommander commented:
/retest

roycaihw commented:
/retest

roycaihw commented on pkg/kubelet/stats/helper.go:
```diff
@@ -493,3 +522,23 @@ func makePodStorageStats(s *statsapi.PodStats, rootFsInfo *cadvisorapiv2.FsInfo,
 	s.EphemeralStorage = calcEphemeralStorage(s.Containers, ephemeralStats, rootFsInfo, logStats, etcHostsStats, isCRIStatsProvider)
 }
+
+func cadvisorPSIToStatsPSI(psi *cadvisorapiv1.PSIStats) *statsapi.PSIStats {
```
Done. I added a unit test `TestCadvisorPSIStructs` in `pkg/kubelet/stats/helper_test.go`

roycaihw commented:
/test pull-kubernetes-node-e2e-containerd-kubelet-psi

roycaihw commented:
[pull-kubernetes-e2e-capz-windows-master](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/130701/pull-kubernetes-e2e-capz-windows-master/1903115812160409600)
> Ginkgo detected a panic while constructing the spec tree

The Windows test failure is unrelated to this PR. It's reported in https://github.com/kubernetes/kubernetes/issues/130960. There is a fix https://github.com/kubernetes/kubernetes/pull/130962 in progress

roycaihw commented:
All tests passed excepted the known [Windows failure](https://github.com/kubernetes/kubernetes/pull/130701#issuecomment-2743961171). This is ready for another round of review @liggitt @haircommander

roycaihw commented:
I also confirm that the [tests passed](https://github.com/kubernetes/kubernetes/pull/131010) with the [above-mentioned Windows test fix](https://github.com/kubernetes/kubernetes/pull/130962). 

/retest

roycaihw commented:
/test pull-kubernetes-node-e2e-containerd-alpha-features

roycaihw commented:
/retest

liggitt commented:
The windows job has been failing since 3/19, looks like - https://prow.k8s.io/job-history/gs/kubernetes-ci-logs/pr-logs/directory/pull-kubernetes-e2e-capz-windows-master?buildId=1902776815471038464

liggitt commented:
mirroring of the cadvisor API type lgtm

haircommander commented:
/lgtm

mbianchidev commented:
/milestone v1.33

roycaihw commented:
/retest
